scanner/src/main/kotlin/LocalScanner.kt

```diff
@@ -300,13 +314,26 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
             downloadResult.dateTime, downloadResult.sourceArtifact, downloadResult.vcsInfo,
             downloadResult.originalVcsInfo
         )
+
+        archiveFiles(downloadResult.downloadDirectory, pkg.id, provenance)
+
         val scanResult = scanPath(downloadResult.downloadDirectory, resultsFile).copy(provenance = provenance)
 
         ScanResultsStorage.storage.add(pkg.id, scanResult)
 
         return scanResult
     }
 
+    private fun archiveFiles(directory: File, id: Identifier, provenance: Provenance) {
+        log.info { "Archiving files for ${id.toCoordinates()}." }
+
+        val provenanceBytes = provenance.toString().toByteArray()
+        val provenanceHash = MessageDigest.getInstance("SHA-1").digest(provenanceBytes).toHexString()
+        val path = "${id.toPath()}/$provenanceHash"
+
+        archiver.save(directory, path)
```

Comment:
Just asking myself aloud: Would this work as expected for consecutive scans of the same package where only the pattern for which files to archive changed? I.e. would any new matching files be added to the directory of the previous scan, but not any not-anymore-matching files be removed? I *do* believe that *is* the case from reading the code, but I want to double-check with you.

Reply:
Depends on what you expect :)
But yes, currently it would add any new files but not delete anything, because the `FileSaver` can only (over)write files.
However, in a new scan of the same package the timestamp of the provenance would be different resulting in a different provenance hash, so a new directory would be used for archiving the files. I considered ignoring the timestamp when building the provenance hash, because while in theory the scanned files for the same provenance could have changed, in reality this is highly unlikely as the provenance contains either the SCM revision or the source artifact hash. I would like to keep it like that for now to be on the safe side, and maybe change it later, but nonetheless, what do you think about this?

Reply:
Yeah, I believe we can leave it like this for now until we eventually see a concrete need to fine-tune the hash generation.

But for a brief moment I was wondering whether we should use something completely different (i.e. not derived from the provenance) as the hash, like the SPDX package verification code for *all* files in the directory (not only the archived ones), as you can't deduce the provenance from the has anyway, and as that would have the advantage of not requiring additional storage for new package versions where the license files did not actually change.

What do you think?

Reply:
We need to look up archived files in other components like the notice reporter based on the provenance of the scan result, so currently only the provenance hash works for this. But the SPDX verification hash is one of the next tasks I will be working on and it will be added to the scan result model, so once it is there we could consider switching to this instead.

Reply:
> But the SPDX verification hash is one of the next tasks I will be working on

Just as a reminder, we already have the (so far unused) `calculatePackageVerificationCode` functions in `SpdxUtils.kt`.

Reply:
Nice, that will come in handy.

Reply:
I made a PR for the SPDX verification code (#1903), what do you think, shall we use this instead for the path? I'm uncertain which is better.

Reply:
That depends on what we decide about https://github.com/heremaps/oss-review-toolkit/pull/1903#discussion_r338413577. IMO it makes only sense to consider the package verification code here if it refers to all files in the package (like you've implemented it for now), not only to the scanned files. But the latter also has its merit.